lib/Zonemaster/Backend/DB.pm

```diff
@@ -241,6 +242,20 @@ sub test_progress {
                 $test_id,
             );
         }
+        elsif ( $progress == 100 ) {
+            $dbh->do(
+                q[
+                    UPDATE test_results
+                    SET progress = ?,
+                        ended_at = ?
+                    WHERE hash_id = ?
+                ],
+                undef,
+                $progress,
+                $self->format_time( time() ),
+                $test_id,
+            );
```

Comment:
I do not feel it is correct that a test with progress=0 could be set directly to 100. What use case is that?

Secondly, there is no check that the update really was done. There should be exactly one record updated.

Reply:
> I do not feel it is correct that a test with progress=0 could be set directly to 100. What use case is that?

Could you elaborate when would this occur?
I'm not sure I see the issue you are raising. However there is still a need to update the `progress` value to `100` once the job is completed and the data is stored. Now that there are 2 tables, this should be done in an independent manner.

> There should be exactly one record updated.

That's why there is a `WHERE hash_id = ?` clause. The hash_id should be unique in the database.

Reply:
I think the code should prevent an error where a test with progress = 0 is updated.

I also think the code should verify that a test really was updated, i.e. that the number of updated records is one not zero.

Reply:
This could be an idea. However, since this is not directly related to this PR, I think it would be better to make such change in an extra PR to not mix things.

Reply:
It looks like it is new code with this PR. But there are more places where the update is done without checking the result.